src/features/game/lib/transforms.ts

```diff
@@ -65,3 +65,65 @@ export function makeGame(farm: any): GameState {
     id: farm.id,
   };
 }
+
+/**
+ * Merges in server side changes into the local state
+ */
+export function updateGame(
+  newGameState: GameState,
+  oldGameState: GameState
+): GameState {
+  if (!newGameState) {
+    return oldGameState;
+  }
+
+  // Only update random number values generated from the server
+  return {
+    ...oldGameState,
+    // Update any random numbers from the server
+    trees: Object.keys(oldGameState.trees).reduce((trees, treeId) => {
```

Comment:
Could we do something like this to avoid the repetition? This could be further simplified if tree had the key `amount` instead of `wood` 😅 . Update resource could then be a generic function.

```
type ResourceData = Record<number, Tree | Rock>;

function objectKeys<Obj>(obj: Obj): (keyof Obj)[] {
  return Object.keys(obj) as (keyof Obj)[];
}

function isTree(resource: Tree | Rock): resource is Tree {
  return (resource as Tree).wood !== undefined;
}

const updateResource = (oldState: ResourceData, newState: ResourceData) => {
  return objectKeys(oldState).reduce((resourceData, itemId) => {
    const id = Number(itemId);
    const resource = oldState[id];

    if (isTree(resource)) {
      return {
        ...resourceData,
        [id]: {
          ...resource,
          wood: (newState[id] as Tree).wood
        }
      }
    }

    return {
      ...resourceData,
      [id]: {
        ...resource,
        amount: (newState[id] as Rock).amount
      }
    }
  }, {})
} 

function updateGame(
  newGameState: GameState,
  oldGameState: GameState
): GameState {
  if (!newGameState) {
    return oldGameState;
  }

  return {
    trees: updateResource(oldGameState.trees, newGameState.trees),
    stones: updateResource(oldGameState.stones, newGameState.stones),
    iron: updateResource(oldGameState.iron, newGameState.iron),
    gold: updateResource(oldGameState.gold, newGameState.gold)
  }
}```

Reply:
In hindsight it would of been good to have similar interfaces for wood + rocks. Right now it would require a migration though which is a bit out of scope 😢 

I've created a reusable function as your suggestion above but only applied it to the rocks 😄